hathor/indexes/rocksdb_timestamp_index.py

```diff
@@ -0,0 +1,151 @@
+# limitations under the License.
+
+from typing import TYPE_CHECKING, Iterator, List, Optional, Tuple
+
+from structlog import get_logger
+
+from hathor.indexes.rocksdb_utils import RocksDBIndexUtils, incr_key
+from hathor.indexes.timestamp_index import RangeIdx, TimestampIndex
+from hathor.transaction import BaseTransaction
+from hathor.util import collect_n, skip_n
+
+if TYPE_CHECKING:  # pragma: no cover
+    import rocksdb
+
+logger = get_logger()
+
+
+class RocksDBTimestampIndex(TimestampIndex, RocksDBIndexUtils):
+    """ Index of transactions sorted by their timestamps.
+
+    This index uses the following key format:
+
+        key = [tx.timestamp][tx.hash]
+              |--4 bytes---||--32b--|
+
+    It works nicely because rocksdb uses a tree sorted by key under the hood.
+    """
+
+    def __init__(self, db: 'rocksdb.DB', *, cf_name: bytes) -> None:
+        RocksDBIndexUtils.__init__(self, db)
+        self.log = logger.new()
+
+        # column family stuff
+        self._cf_name = cf_name
+        self._cf = self._fresh_cf(self._cf_name)
+
+    def _to_key(self, timestamp: int, tx_hash: Optional[bytes] = None) -> bytes:
+        """Make a key for a timestamp and optionally tx_hash, the key represents the membership itself."""
+        import struct
+        key = bytearray()
+        key.extend(struct.pack('>I', timestamp))
+        if tx_hash is None:
+            assert len(key) == 4
+            return bytes(key)
+        assert len(tx_hash) == 32
+        key.extend(tx_hash)
+        assert len(key) == 4 + 32
+        return bytes(key)
+
+    def _from_key(self, key: bytes) -> Tuple[int, bytes]:
+        """Parse a key on the column-family."""
+        import struct
+        assert len(key) == 4 + 32
+        timestamp: int
+        (timestamp,) = struct.unpack('>I', key[:4])
+        tx_hash = key[4:]
+        assert len(tx_hash) == 32
+        return timestamp, tx_hash
+
+    def add_tx(self, tx: BaseTransaction) -> bool:
+        key = self._to_key(tx.timestamp, tx.hash)
+        is_new = self._db.get((self._cf, key)) is None
+        if is_new:
+            self._db.put((self._cf, key), b'')
+        return is_new
+
+    def del_tx(self, tx: BaseTransaction) -> None:
+        key = self._to_key(tx.timestamp, tx.hash)
+        self._db.delete((self._cf, key))
+
+    def _iter(self, from_timestamp: Optional[int] = None, from_tx: Optional[bytes] = None,
+              *, reverse: bool = False) -> Iterator[Tuple[int, bytes]]:
+        """ Iterate over all transactions of a token, by default from oldest to newest.
+        """
+        if from_timestamp is None and from_tx is not None:
+            raise ValueError('from_tx needs from_timestamp, but it is None')
+        it = self._db.iterkeys(self._cf)
+        if reverse:
+            it = reversed(it)
+            if from_timestamp is None:
+                self.log.debug('seek to last')
+                it.seek_to_last()
```

Comment:
This was a bit confusing to me. If you already reversed the iterkeys above (`it = reversed(it)`) why do you need to `seek_to_last`? I thought you would need to get the first elements here, no?

Reply:
As far as I can tell, "first" and "last" are always relative to the database, and don't change when the iterator is reversed. Or rather, reverting the iterator only changes the direction 'next()' will go.